src/rules.jl

```diff
@@ -96,3 +96,110 @@ true
 See also: [`frule`](@ref), [`@scalar_rule`](@ref)
 """
 rrule(::Any, ::Vararg{Any}; kwargs...) = nothing
+
+
+#######################################################################
+# Infastructure to support generating overloads from rules.
+
+const NEW_RRULE_HOOKS = Function[]
+const NEW_FRULE_HOOKS = Function[]
+_hook_list(::typeof(rrule)) = NEW_RRULE_HOOKS
+_hook_list(::typeof(frule)) = NEW_FRULE_HOOKS
+
+"""
+    on_new_rule(sig->eval...), frule | rrule)
+"""
+function on_new_rule(hook_fun, rule_kind)
+    # get all the existing rules
+    ret = map(_rule_list(rule_kind)) do method
+        sig = _primal_sig(rule_kind, method)
+        _safe_hook_fun(hook_fun, sig)
+    end
+
+    # register hook for new rules
+    push!(_hook_list(rule_kind), hook_fun)
+    return ret
+end
+
+function __init__()
+    push!(Base.package_callbacks, pkgid -> refresh_rules())
+    push!(Base.include_callbacks, (mod, filename) -> refresh_rules())
+end
+
+
+"""
+    _rule_list(frule | rrule)
+
+Returns a list of all the methods of the currently defined rules of the given kind.
+Excluding the fallback rule that returns `nothing` for every input.
+"""
+_rule_list(rule_kind) = (m for m in methods(rule_kind) if m.module != @__MODULE__)
+# ^ The fallback rules are the only rules defined in ChainRules core  so that is how we skip them.
+
+
+
+const LAST_REFRESH_RRULE = Ref(0)
+const LAST_REFRESH_FRULE = Ref(0)
+last_refresh(::typeof(frule)) = LAST_REFRESH_FRULE
+last_refresh(::typeof(rrule)) = LAST_REFRESH_RRULE
+
+"""
+    refresh_rules(); refresh_rules(frule); refresh_rules(rrule)
+
+This triggers all [`on_new_rule`](@ref) hooks to run on any newly defined rules.
+It is *automatically* run when ever a package is loaded, or a file is `include`d.
+It can also be manually called to run it directly, for example if a rule was defined
+in the REPL or with-in the same file as the AD function.
+"""
+refresh_rules() = (refresh_rules(frule); refresh_rules(rrule))
+function refresh_rules(rule_kind)
+    already_done_world_age = last_refresh(rule_kind)[]
```

Comment:
should we check here if there are any hooks regisistered, and if not bail out early?
So that if someone has something that uses ChainRulesCore, but doesn't use any overloadinging AD package, then nothing is added?